src/particles.js

```diff
@@ -0,0 +1,167 @@
+import * as util from "./util";
+import Layer from "./layer";
+
+import particleUpdateVert from "./shaders/particle-update.vert.glsl";
+import particleUpdateFrag from "./shaders/particle-update.frag.glsl";
+import particleDrawVert from "./shaders/particle-draw.vert.glsl";
+import particleDrawFrag from "./shaders/particle-draw.frag.glsl";
+
+class Particles extends Layer {
+  constructor(options) {
+    this.propertySpec = {
+      "particle-color": {
+        type: "color",
+        default: "white",
+        expression: {
+          interpolated: true,
+          parameters: ["zoom", "feature"]
+        },
+        "property-type": "data-driven"
+      },
+      "particle-speed": {
+        type: "number",
+        minimum: 0,
+        default: 0.75,
+        transition: true,
+        expression: {
+          interpolated: true,
+          parameters: ["zoom"]
+        },
+        "property-type": "data-constant"
+      }
+    };
+    super(options);
+    this.dropRate = 0.003; // how often the particles move to a random place
+    this.dropRateBump = 0.01; // drop rate increase relative to individual particle speed
+    this._numParticles = 65536;
+  }
+
+  setParticleColor(expr) {
+    this.buildColorRamp(expr);
+  }
+
+  initializeParticles(gl, count) {
+    const particleRes = (this.particleStateResolution = Math.ceil(
+      Math.sqrt(count)
+    ));
+    this._numParticles = particleRes * particleRes;
+
+    const particleState = new Uint8Array(this._numParticles * 4);
+    for (let i = 0; i < particleState.length; i++) {
+      particleState[i] = Math.floor(Math.random() * 256); // randomize the initial particle positions
+    }
+    // textures to hold the particle state for the current and the next frame
+    this.particleStateTexture0 = util.createTexture(
+      gl,
+      gl.NEAREST,
+      particleState,
+      particleRes,
+      particleRes
+    );
+    this.particleStateTexture1 = util.createTexture(
+      gl,
+      gl.NEAREST,
+      particleState,
+      particleRes,
+      particleRes
+    );
+
+    const particleIndices = new Float32Array(this._numParticles);
+    for (let i = 0; i < this._numParticles; i++) particleIndices[i] = i;
+    this.particleIndexBuffer = util.createBuffer(gl, particleIndices);
+  }
+
+  initialize(map, gl) {
+    this.updateProgram = util.createProgram(
+      gl,
+      particleUpdateVert,
+      particleUpdateFrag
+    );
+
+    this.drawProgram = util.createProgram(
+      gl,
+      particleDrawVert,
+      particleDrawFrag
+    );
+
+    this.framebuffer = gl.createFramebuffer();
+
+    this.quadBuffer = util.createBuffer(
+      gl,
+      new Float32Array([0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1])
+    );
+
+    this.initializeParticles(gl, this._numParticles);
+  }
+
+  prerender(gl, matrix) {
+    if (this.windData) this.update(gl, matrix);
+  }
+
+  update(gl, matrix) {
```

Comment:
Some explanation of what this code does would be good. Not knowing WebGL, I guess it updates particle state etc but that is the level of my understanding. If anyone else is to help maintain this code, some contextual hints would be useful, it the kind of thing we will visit rarely, so docs are important.